Goto contractor: support for multiple operators.

Mohannad-Aldughaim opened the pull request:
Hello everyone,

This is an update to the --goto-contractor option. This update includes support for operators &&, ||, ==, !=, and multiple asserts. Regression tests show the new features. Instead of creating one constraint and one contractor, the algorithm will create multiple contractors and save their location in the program. This will enable the algorithm to take the intersection and union of multiple properties and create their complement. The saved location helps get the properties to a given point in the program. 

I am happy to answer any questions you have.

Regards,
-Mohannad

Mohannad-Aldughaim committed:
-Intermediate commit. Includes support for operators: ==, !=, &&, ||, !
-Implemented create contractors from expressions.
-Implemented two classes to manage contractors.

Mohannad-Aldughaim committed:
-Intermediate commit. Includes support for operators: ==, !=, &&, ||, !
-Further Implementation two classes to manage contractors.
  *Can add contractors
  *get complement contractors

Mohannad-Aldughaim committed:
-Bug fixes and tidy
-Added regression tests for new features

Mohannad-Aldughaim committed:
-Code-style

Mohannad-Aldughaim committed:
-tidy + code-style

Mohannad-Aldughaim committed:
-fixed regression tests. (missing a #include <assert.h>)

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -129,11 +132,21 @@ void goto_contractort::insert_assume(goto_functionst goto_functions)
       {
-        //only update upperbound
-        auto ub = create_value_expr(var.second.getInterval().ub(), int_type2());
+        auto lb =
```
Can you please add some comments?

Mohannad-Aldughaim replied:
Previously, we only reduced the upper bound to avoid false negatives. I'm testing with both bounds this time and waiting on the results.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -142,24 +155,18 @@ void goto_contractort::insert_assume(goto_functionst goto_functions)
-
-  ibex::NumConstraint c2(c.f, complement);
-  oss << "\n\t- Complement: " << c2;
+  ibex::CtcFixPoint c_out(*contractor->get_outer());
```
Can you relate this to the operators we describe in our paper? For example, add some short sentences to explain the inner and outer contractor.

Mohannad-Aldughaim replied:
done

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -142,24 +155,18 @@ void goto_contractort::insert_assume(goto_functionst goto_functions)
 
   domains = map.create_interval_vector();
 
+  oss << "Contractors applied:";
```
Shouldn't we use the fmt library interface to output messages?

Mohannad-Aldughaim replied:
I am using fmt. The `std::ostringstream oss;` is just to format the string.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -171,37 +178,77 @@ void goto_contractort::contractor()
+  auto base_object = get_base_object(expr);
+
+  if(
+    (is_comp_expr(base_object) && !is_notequal2t(base_object)) ||
```
This if-condition looks weird to me. Can you add some comments?

Mohannad-Aldughaim replied:
Added to the documentation of this function.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -171,37 +178,77 @@ void goto_contractort::contractor()
+      break;
+    }
+    case expr2t::expr_ids::not_id:
+      //TODO: implement "not" flag  and process.
```
You could print a message here.

Mohannad-Aldughaim replied:
done.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -171,37 +178,77 @@ void goto_contractort::contractor()
+      //TODO: implement "not" flag  and process.
+      break;
+    default:
+      //unsupported. you shouldn't be here.
```
You could print a message here.

Mohannad-Aldughaim replied:
done.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -171,37 +178,77 @@ void goto_contractort::contractor()
     std::ostringstream oss;
     oss << get_expr_id(expr);
-    oss << "Expression is complex, skipping this assert.\n";
+    oss << " Expression is complex, skipping this assert.\n";
```
Can you define the ``complex`` term?

Mohannad-Aldughaim replied:
In Ibex, the expressions are structured more strictly than c. On the top level, we have logical operators such as && and ||. Then the comparison operators such as > and <. For example, we can have `(x > y || x < y)` and parse successfully. However, if the expression is valid in c `( x == y > 0 )`, it cannot be converted to ibex without reasoning. That's why we call them complex expressions.

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -171,37 +178,77 @@ void goto_contractort::contractor()
   }
+  if(is_not2t(expr))
+  {
+    //TODO: implement "not" flag  and process.
```
You could print a message here.

Mohannad-Aldughaim replied:
done

lucasccordeiro commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -307,6 +367,13 @@ goto_contractort::create_function_from_expr2t(irep_container<expr2t> expr)
     f = new ibex::Function(*vars, c);
     break;
   }
+  case expr2t::expr_ids::constant_floatbv_id:
```
Do we really support float?

Mohannad-Aldughaim replied:
We don't yet. removed. I was trying to handle floats in this update, but I faced some problems with it.

Mohannad-Aldughaim committed:
-Removed partial support for floats.
-Added documentation to new functions.
-Added parse_error to print on the screen where it stopped parsing.

Mohannad-Aldughaim committed:
-Code-style

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -41,6 +41,210 @@ class vart
+  ibex::Ctc *get_complement_contractor(ibex::Ctc *c)
+  {
+    ibex::Ctc *complement_contractor = nullptr;
+    auto type = boost::core::demangle(typeid(*c).name());
```
doesn't ibex have a way of getting some kind of type id? IIRC, typeid.name() may not be portable.

Mohannad-Aldughaim replied:
Unfortunately, that's the only way I found to determine the type in ibex.

mikhailramalho replied:
I just had a look at the API and all these objects inherit from ctc, so you can use the dynamic_cast that you already implemented to convert them safety.

It will look something like: 
```
if(auto ctc_compo = dynamic_cast<ibex::CtcCompo *>(c)) {
      ibex::Array<ibex::Ctc> list_of_contractors;
      for(auto &it : ctc_compo->list)
        list_of_contractors.add(*get_complement_contractor(&it));
      return new ibex::CtcUnion(list_of_contractors);
} else if (auto ctc_union = dynamic_cast<ibex::CtcUnion *>(c)) {
      auto ctc_union = dynamic_cast<ibex::CtcUnion *>(c);
      ibex::Array<ibex::Ctc> list_of_contractors;
      for(auto &it : ctc_union->list)
        list_of_contractors.add(*get_complement_contractor(&it));
      return new ibex::CtcCompo(list_of_contractors);
} // etc

      std::ostringstream oss;
      oss << "Type: " << type;
      log_debug("{}", oss.str());

     return nullptr;
```

Note that you don't need `complement_contractor` at all here.

Mohannad-Aldughaim replied:
Done. What I had in mind when writing it was something like java's `instanceof`. Thank you!

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -41,6 +41,210 @@ class vart
+    return c;
+  }
+
+  Contractor *get_contractors(void)
```
no need for K&R style.

Mohannad-Aldughaim replied:
Not sure what you mean. is it the spacing between the functions?

mikhailramalho replied:
it's the `function_name(void)`. Just leave the it empty, i.e., `function_name()`.

Mohannad-Aldughaim replied:
done.

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -153,23 +355,21 @@ class goto_contractort : public goto_functions_algorithm
       {
-        log_status(
-          "Constraint expression not supported. Aborting goto-contractor");
+        log_status("Aborting goto-contractor, No Contractors were created.");
```
I suggest avoiding using  "abort" in status messages. Maybe esbmc can print something like "Contractors: No supported expression found".

Mohannad-Aldughaim replied:
done

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -50,7 +254,7 @@ class vart
 {
 public:
-  static constexpr int MAX_VAR = 10;
+  static constexpr int MAX_VAR = 26;
```
I think I've asked this before but do we really need this MAX_VAR constant?

Mohannad-Aldughaim replied:
removed

mikhailramalho replied:
You removed it but used the magic number in the code directly. Please, use the MAX_VAR constant (or any other constant name FWIW) instead of magic numbers.

My question is why do we need to limit the max number of constants, doesn't the code work with more than 26 variables?

Mohannad-Aldughaim replied:
I agree with you. It should not be arbitrary. I am still working on this. Because I can't change the size of the vector, I need to recreate it each time we find a new variable. The ibex documentation did not state any limits on the number of variables in a vector.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -69,26 +72,27 @@ void goto_contractort::parse_intervals(expr2tc expr)
+  if(is_constant_int2t(side2))
+    value = to_constant_int2t(side2).as_long() * (neg ? -1 : 1);
+  else
+    return;
```
Can you revert this to the previous early return code?

Mohannad-Aldughaim replied:
done

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -171,35 +180,74 @@ void goto_contractort::contractor()
+  ibex::Ctc *contractor = nullptr;
+  auto base_object = get_base_object(expr);
+
+  if(
```
can you rewrite this as an early return? i.e, if it's none of the expect exprt, then return.

mikhailramalho replied:
or better still: you can use the is_unsupported_exprt you created.

Mohannad-Aldughaim replied:
Created a new one for contractors. Because some are supported on the contractor level but not on the constraint level.

Mohannad-Aldughaim replied:
I made a new function that does the same as the condition of the `if` statement. its called `is_constraint_operator()`

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -316,13 +378,25 @@ goto_contractort::create_function_from_expr2t(irep_container<expr2t> expr)
 
 int goto_contractort::create_variable_from_expr2t(irep_container<expr2t> expr)
```
use expr2tc instead of irep_container<expr2t>

Mohannad-Aldughaim replied:
Done

mikhailramalho replied:
I meant in the function parameter

Mohannad-Aldughaim replied:
done. My bad I thought I changed it.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -316,13 +378,25 @@ goto_contractort::create_function_from_expr2t(irep_container<expr2t> expr)
   {
-    index = map.add_var(var_name, to_symbol2t(expr));
+    std::string var_name = to_symbol2t(expr).get_symbol_name();
+    int index = map.find(var_name);
```
you can probably just use add, instead of find + add.

Mohannad-Aldughaim replied:
done.

mikhailramalho reviewed (changes requested):
Can you also check if the expressions are not leaking? I see a lot of `new` calls but no `delete`

lucasccordeiro commented:
@Mohannad-Aldughaim: Can I ask you for the status of this PR?

Mohannad-Aldughaim committed:
-Bug fixes and tidy.
-Addressed comments from PR852

Mohannad-Aldughaim commented:
Many thanks for the comments and sorry for the delay.
In preparation for SV-Comp, May I ask you to review and merge this pull request? 

@mikhailramalho :
>  Can you also check if the expressions are not leaking? I see a lot of new calls but no delete.

It seems to me that ibex uses smart pointers. when I try to delete the object it raises an exception.

Mohannad-Aldughaim committed:
-Lint

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -69,26 +72,25 @@ void goto_contractort::parse_intervals(expr2tc expr)
     return;
 
-  value = to_constant_int2t(side2).as_long() * (neg ? -1 : 1);
+  value = to_constant_int2t(side2).as_long() * (neg ? -1 : 1) + (neg ? -1 : 1);
```
What does this do exactly?

Mohannad-Aldughaim replied:
When I use the `--interval-analysis` option it will insert assumes into the code. These assumes do not hold in some cases because of a difference of 1. For example:
```
int x=0;
int y=0;
while( y < 1000 )
{
   assume(y<=999) //inserted by interval analysis option
   x=x+y;
   y++;
}
assert(x >=y)
```
Here the interval analysis option says the upper bound is 999. However, when we reach the assert it will be 1000. that's why I added/subtracted 1 in that statement. It improved the results for some benchmarks.

I am currently working on a script that will extract accurate intervals using Frama-C.   This is just temporary.

mikhailramalho replied:
I see... The interval analysis is missing the loop termination condition that it's added but the k-induction transformations.

You can probably move that from the k-induction algorithm to a separate transformation step in a separate PR, so you can remove this code. WDYT @lucasccordeiro?

lucasccordeiro replied:
Yes, I agree that it could be a separate PR.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -171,35 +169,89 @@ void goto_contractort::contractor()
-  default:
-    log_status("cant process equal");
-    break;
+    auto cons = create_constraint_from_expr2t(base_object);
```
is it's unsupported, shouldn't we just return nullptr here?

Mohannad-Aldughaim replied:
I changed the name of the function to show its purpose. Because if we encounter these operators we will create constraints, not contractors.

mikhailramalho commented on src/goto-programs/goto_contractor.cpp:
```diff
@@ -225,28 +278,32 @@ goto_contractort::create_constraint_from_expr2t(expr2tc expr)
-         is_notequal2t(e) || is_equality2t(e) || is_not2t(e) ||
-         is_modulus2t(e) || is_or2t(e) || is_and2t(e);
+         is_notequal2t(e) || is_not2t(e) || is_modulus2t(e) || is_or2t(e) ||
+         is_and2t(e);
```
I think you've implemented the support for `and` and `notequal`, right?

Mohannad-Aldughaim replied:
Correct. I changed the function name. sorry for the confusion.

Mohannad-Aldughaim committed:
-Addressed Comments on PR852.
-Changed function name from is_unsuppoted_operator_in_contractor to is_constraint_operator.
-Changed Dynamic_cast in get_complement_contractor.

mikhailramalho commented on src/goto-programs/goto_contractor.h:
```diff
@@ -41,6 +41,198 @@ class vart
+  {
+    std::ostringstream oss;
+    auto type = boost::core::demangle(typeid(*c).name());
+    if(type == "ibex::CtcCompo" || type == "ibex::CtcUnion")
```
you don't need the typeid here, just use dynamic_cast and check if the type != nullptr

Mohannad-Aldughaim replied:
done

mikhailramalho reviewed (approved):
Thanks for addressing my comments.

Please just remove the remaining typeid and this should be good to go

Mohannad-Aldughaim committed:
-Addressed Comments on PR852.
-Removed the remaining typeid.
